components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -267,6 +265,79 @@ Note that if there is association between a federated user and local user () 'As
         return userAllowedScopes;
     }
 
+    private Set<Scope> getApplicationAllowedScopes(String clientId, String[] requestedScopes, String tenantDomain) {
+
+        Set<Scope> userAllowedScopes = new HashSet<>();
+
+        try {
+            if (requestedScopes == null) {
+                return new HashSet<>();
+            }
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+            Set<Scope> allScopes = getScopesOfPermissionType(tenantId);
+            if (ArrayUtils.contains(requestedScopes, SYSTEM_SCOPE)) {
+                requestedScopes = getScopeNames(allScopes);
+            } else {
+                // filter out the internal scopes
+                requestedScopes = Oauth2ScopeUtils.getRequestedScopes(requestedScopes);
+            }
+            Set<String> requestedScopesSet = new HashSet<>(Arrays.asList(requestedScopes));
+
+            startTenantFlow(tenantDomain, tenantId);
```

Comment:
start tenant flow should be the first statement of the try block, to avoid end tenant flow getting invoked in finally block when an exception occur before reaching here

Reply:
Fixed. But please verify again